### ISSUE TITLE
Brainstorming: Drivers

### ISSUE THREAD

**Issue by skyem123** [XeonSquared/PsychOS#1]
Ideally it would have drivers as individual processes using events to communicate.
However that would probably be horrendusly inefficent especially for commonly used components (such as graphics), it would be more practical for simpler things, such as redstone and disk drive (not FS or raw block devices!).

Maybe the "component" driver (or "HAL"?) could run divers inside of it, using sandboxing to ensure they are isolated from the rest of the system, except for the components the HAL gives it, the core Lua APIs, and a way to expose it's own APIs to the rest of the system (events? just normal APIs? both?).

This is a very vague idea and probably won't be practical, but it would be interesting.

**Comment by XeonSquared** [comment]
I was thinking it would be more efficient to have the kernel wrap the component API, perhaps with a table containing all components, which new ones could be added to for virtual devices. Real hardware would be direct, emulated hardware would seem direct.

**Comment by skyem123** [comment]
True... Though the raw component API is not really fun to use, though I guess you could patch that.

**Comment by XeonSquared** [comment]
Hopefully reasonably completed with 5300a21a3752d6af4bed0ab8a202398e5eafb2e5. Gonna leave this open in case there's more ideas for improvement.
